libs/autotagging-commons/src/main/java/org/opensearch/rule/service/IndexStoredRulePersistenceService.java

```diff
@@ -0,0 +1,135 @@
+        this.indexName = indexName;
+        this.client = client;
+        this.featureType = featureType;
+        this.maxRulesPerPage = maxRulesPerPage;
+    }
+
+    /**
+     * Entry point for the get rule api logic in persistence service.
+     * @param getRuleRequest the getRuleRequest to process.
+     * @param listener the listener for GetRuleResponse.
+     */
+    public void getRule(GetRuleRequest getRuleRequest, ActionListener<GetRuleResponse> listener) {
+        getRuleFromIndex(getRuleRequest.getId(), getRuleRequest.getAttributeFilters(), getRuleRequest.getSearchAfter(), listener);
+    }
+
+    /**
+     * Get rules from index. If id is provided, we only get a single rule.
+     * Otherwise, we get all rules that satisfy the attributeFilters.
+     * @param id - The id of the rule to get.
+     * @param attributeFilters - A map containing the attributes that user want to filter on
+     * @param searchAfter - The sort values from the last document of the previous page, used for pagination
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    public void getRuleFromIndex(
+        String id,
+        Map<Attribute, Set<String>> attributeFilters,
+        String searchAfter,
+        ActionListener<GetRuleResponse> listener
+    ) {
+        // Stash the current thread context when interacting with system index to perform
+        // operations as the system itself, bypassing authorization checks. This ensures that
+        // actions within this block are trusted and executed with system-level privileges.
+        try (ThreadContext.StoredContext context = getContext()) {
+            BoolQueryBuilder boolQuery = IndexStoredRuleUtils.buildGetRuleQuery(id, attributeFilters, featureType);
+            SearchRequestBuilder searchRequest = client.prepareSearch(indexName).setQuery(boolQuery).setSize(maxRulesPerPage);
+            if (searchAfter != null) {
+                searchRequest.addSort(_ID_STRING, SortOrder.ASC).searchAfter(new Object[] { searchAfter });
+            }
+            searchRequest.execute(ActionListener.wrap(searchResponse -> handleGetRuleResponse(id, searchResponse, listener), e -> {
+                logger.error("Failed to fetch all rules: {}", e.getMessage());
+                listener.onFailure(e);
+            }));
+        }
+    }
+
+    /**
+     * Process searchResponse from index and send a GetRuleResponse
+     * @param searchResponse - Response received from index
+     * @param listener - ActionListener for GetRuleResponse
+     */
+    void handleGetRuleResponse(String id, SearchResponse searchResponse, ActionListener<GetRuleResponse> listener) {
+        List<SearchHit> hits = Arrays.asList(searchResponse.getHits().getHits());
+        if (id != null && hits.isEmpty()) {
+            logger.error("Rule with ID " + id + " not found.");
+            listener.onFailure(new ResourceNotFoundException("Rule with ID " + id + " doesn't exist in the .rules index."));
```

Comment:
I don't think we should be propagating exception messages as `doesn't exist in .rules index`, as user should be agnostic of the underlying storage mechanism for rules.